plugins/legion-go2/fu-legion-go2-device.c

```diff
@@ -0,0 +1,671 @@
+			return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+static gchar*
+fu_legion_go2_device_get_version(FuLegionGo2Device *self, guint8 id, GError **error)
+{
+	g_autoptr(FuStructLegionGo2NormalCmd) versionCmd = fu_struct_legion_go2_normal_cmd_new();
+	guint8 content[] = { 
+		0x01
+	};
+	fu_struct_legion_go2_normal_cmd_set_report_id(versionCmd, 5);
+	fu_struct_legion_go2_normal_cmd_set_length(versionCmd, sizeof(content) + 4);
+	fu_struct_legion_go2_normal_cmd_set_main_id(versionCmd, 0x79);
+	fu_struct_legion_go2_normal_cmd_set_sub_id(versionCmd, 0x01);
+	fu_struct_legion_go2_normal_cmd_set_device_id(versionCmd, id);
+	fu_struct_legion_go2_normal_cmd_set_data(versionCmd, content, sizeof(content), error);
+
+	if (!fu_udev_device_write(FU_UDEV_DEVICE(self),
+				  versionCmd->buf->data,
+				  versionCmd->buf->len,
+				  FU_LEGION_GO2_DEVICE_IO_TIMEOUT,
+				  FU_IO_CHANNEL_FLAG_NONE,
+				  error)) {
+		g_message("send version command failed: ");
+		return NULL;
+	}
+
+	g_autoptr(GByteArray) response = g_byte_array_sized_new(FU_LEGION_GO2_DEVICE_FW_REPORT_LENGTH);
+	g_byte_array_set_size(response, FU_LEGION_GO2_DEVICE_FW_REPORT_LENGTH);
+	if (!fu_legion_go2_device_read_response_ex(self, 0x79, 0x01, id, response))
+	{
+		g_message("read version response failed: ");
+		return NULL;
+	}
+
+	gchar* version = g_strdup_printf("%x%02x%02x%02x", 
+					  response->data[13], 
+					  response->data[14], 
+					  response->data[15], 
+					  response->data[16]);
+	g_message("device (%d) version: %s", id, version);
+	
+	return version;
+}
+
+static guint16
+fu_legion_go2_device_calc_crc16(GByteArray* buffer)
+{
+	const guint16 POLY = 0x1021;
+	guint16 crc = 0x0000;
+
+	for (guint i = 0; i < buffer->len; i++)
+	{
+		crc ^= (guint16)(buffer->data[i] << 8);
+		for (int j = 0; j < 8; j++)
+		{
+			if ((crc & 0x8000) != 0)
+			{
+				crc = (guint16)((crc << 1) ^ POLY);
+			}
+			else
+			{
+				crc = (guint16)(crc << 1);
+			}
+		}
+	}
+
+	return crc;
+}
+
+static guint8 fu_legion_go2_device_get_id(GByteArray* buffer)
+{
+	if (buffer->len < FU_LEGION_GO2_DEVICE_FW_SIGNED_LENGTH + FU_LEGION_GO2_DEVICE_FW_ID_LENGTH)
+	{
+		return 0;
+	}
+
+	guint offset = buffer->len - FU_LEGION_GO2_DEVICE_FW_SIGNED_LENGTH - FU_LEGION_GO2_DEVICE_FW_ID_LENGTH;
+	for (guint i = 0; i < FU_LEGION_GO2_DEVICE_FW_ID_LENGTH; ++i)
+	{
+		guint8 id = buffer->data[offset + i];
+		if (id == 2 || id == 7 || id == 8)
+		{
+			return id;
+		}
+	}
+
+	return 0;
+}
+
+static gboolean
+fu_legion_go2_device_probe(FuDevice *device, GError **error)
+{
+	guint16 vid = fu_device_get_vid(FU_DEVICE(device));
+	guint16 pid = fu_device_get_pid(FU_DEVICE(device));
+
+	if (vid == FU_LEGION_GO2_DEVICE_VID && 
+	    pid >= FU_LEGION_GO2_DEVICE_PID_BEGIN && 
+	    pid <= FU_LEGION_GO2_DEVICE_PID_END) {
+		fu_device_set_name(device, "Legion Go2 MCU");
+		fu_device_set_vendor(device, "Lenovo");
+		fu_device_add_flag(device, FWUPD_DEVICE_FLAG_UPDATABLE);
+		fu_device_add_protocol(device, "com.legion.go2");
```

Comment:
these need to go into _init() -- the probe won't be called if the VID/PID isn't listed in the quirk file.